### PR TITLE
SDL: Select audio device

### LINKED ISSUE #1838: SDL2: Selecting an audio device
We currently call [`SDL_OpenAudioDevice`](https://wiki.libsdl.org/SDL_OpenAudioDevice) with a `nullptr` as the first argument because this does (hopefully sensible) automatic selection of the most appropriate audio device. [A user reports](https://gbatemp.net/threads/citra-new-3ds-emulator.365154/page-210#post-6369199) this not to be the case.

SDL does let us list the available audio devices with [`SDL_GetAudioDeviceName`](https://wiki.libsdl.org/SDL_GetAudioDeviceName); this could be a configuration option for a user to override the automatic selection.

### PR DESCRIPTION
**Pull request by Kloen**
Picked up: #1952 because it seems it wasn't going to be finished anytime soon.

This should closes #1838.

Mandatory screenshots:
Music + Citra, same output device
image

Music + Citra diferent output device
image

### CODE REVIEW

File: src/audio_core/sdl2_sink.cpp [hunk 1]
```diff
@@ -42,10 +43,24 @@ SDL2Sink::SDL2Sink() : impl(std::make_unique<Impl>()) {
...
+    if (device_count >= 1 && Settings::values.audio_device_id != "auto" &&
+        !Settings::values.audio_device_id.empty()) {
+        device = Settings::values.audio_device_id.c_str();
+    }
+
+    impl->audio_device_id = SDL_OpenAudioDevice(device, false, &desired_audiospec,
+                                                &obtained_audiospec, SDL_AUDIO_ALLOW_ANY_CHANGE);
     if (impl->audio_device_id <= 0) {
-        LOG_CRITICAL(Audio_Sink, "SDL_OpenAudioDevice failed with: %s", SDL_GetError());
+        LOG_CRITICAL(Audio_Sink, "SDL_OpenAudioDevice failed with code %d for device \"auto\"",
```
Comment (Kloen) [inline src/audio_core/sdl2_sink.cpp]: Note to self: This `LOG_CRITICAL` needs some atention. The message is wrong after the original PR if's changes.

**Comment by MerryMage** [comment]
Would prefer it if `std::vector<std::string>*` → `std::vector<std::string>`.

`NullSink::GetDeviceList()` would then `return {};`.

Other than that looks fine.

**Comment by Kloen** [comment]
Cleaned up commit history. 

Done the point 1 of the @MerryMage comments, I'll do the point 2 too, so just wait for it, but I wanted to push the cleanup already because the commit history was driving me crazy.

File: src/audio_core/sink.h [hunk 2]
```diff
@@ -31,6 +31,15 @@ class Sink {
...
     virtual std::size_t SamplesInQueue() const = 0;
+
+    /**
+     * Sets the desired output device.
+     * @paran device_id Id of the desired device.
+     */
+    virtual void SetDevice(int device_id) = 0;
+
+    /// Returns the list of available devices.
+    virtual std::vector<std::string> GetDeviceList() = 0;
```
Comment (Subv) [inline src/audio_core/sink.h]: This function can be const

File: src/audio_core/sdl2_sink.cpp [hunk 3]
```diff
@@ -96,6 +114,10 @@ size_t SDL2Sink::SamplesInQueue() const {
     return total_size;
 }
 
+void SDL2Sink::SetDevice(int _device_id) {
```
Comment (Subv) [inline src/audio_core/sdl2_sink.cpp]: Why is this _device_id instead of device_id?

File: src/audio_core/sdl2_sink.h [hunk 4]
```diff
@@ -21,9 +21,14 @@ class SDL2Sink final : public Sink {
 
     size_t SamplesInQueue() const override;
 
+    std::vector<std::string> GetDeviceList() override;
+    void SetDevice(int _device_id);
```
Comment (Subv) [inline src/audio_core/sdl2_sink.h]: same

File: src/citra_qt/configure_audio.cpp [hunk 5]
```diff
@@ -33,12 +38,49 @@ void ConfigureAudio::setConfiguration() {
...
+    if (sink_index > 0) {
+
+        // Case where the sink is pointed to a directly known sink device (NullSink, SDL2Sink).
+        // The first index (0) should be "auto", which is not in the vector list (-1).
+        quick_sink_populate_device = AudioCore::g_sink_details[sink_index - 1];
+    }
+    auto iter = quick_sink_populate_device.factory();
+
+    std::vector<std::string> device_list = iter->GetDeviceList();
+    for (const auto device : device_list) {
```
Comment (Subv) [inline src/citra_qt/configure_audio.cpp]: const auto&

File: src/citra_qt/configure_audio.h [hunk 6]
```diff
@@ -20,6 +20,9 @@ class ConfigureAudio : public QWidget {
 
     void applyConfiguration();
 
+public slots:
+    void updateAudioDevices(int sinkIndex);
```
Comment (Kloen) [inline src/citra_qt/configure_audio.h]: `sink_index`

File: src/audio_core/sdl2_sink.cpp (hunk 3 above)
Comment (Kloen) [inline src/audio_core/sdl2_sink.cpp]: I really don't know, original code was this way. Changing it...

File: src/audio_core/sdl2_sink.cpp [hunk 7]
```diff
@@ -69,6 +83,10 @@ unsigned int SDL2Sink::GetNativeSampleRate() const {
     return impl->sample_rate;
 }
 
+std::vector<std::string> SDL2Sink::GetDeviceList() {
```
Comment (Subv) [inline src/audio_core/sdl2_sink.cpp]: missing the const here

**Comment by Kloen** [comment]
All done.

File: src/citra_qt/configure_audio.cpp [hunk 8]
```diff
@@ -33,12 +38,42 @@ void ConfigureAudio::setConfiguration() {
...
     Settings::Apply();
 }
+
+void ConfigureAudio::updateAudioDevices(int sink_index) {
+
+    ui->audio_device_combo_box->clear();
+    ui->audio_device_combo_box->addItem("auto");
+
+    std::vector<std::string> device_list =
+        AudioCore::GetSinkDetails(ui->output_sink_combo_box->itemText(sink_index).toStdString())
```
Comment (Subv) [inline src/citra_qt/configure_audio.cpp]: IMO, it'd be better to use a variable to hold the string and shorten this line.

File: src/citra_qt/configure_audio.cpp [hunk 9]
```diff
@@ -33,12 +38,42 @@ void ConfigureAudio::setConfiguration() {
...
             .toStdString();
     Settings::values.enable_audio_stretching = ui->toggle_audio_stretching->isChecked();
+    Settings::values.audio_device_id =
+        ui->audio_device_combo_box->itemText(ui->audio_device_combo_box->currentIndex())
+            .toStdString();
     Settings::Apply();
 }
+
+void ConfigureAudio::updateAudioDevices(int sink_index) {
+
```
Comment (Subv) [inline src/citra_qt/configure_audio.cpp]: drop the extra blank line

**Comment by Kloen** [comment]
@Subv np, addressed.

**Pull request merged**